src/TuxState.js

```diff
@@ -0,0 +1,650 @@
+var invariantArgCheck = function (input) {
+  if (!input || typeof(input) !== 'object' || Array.isArray(input)) {
+    invariant(!input, 'This function requires an object as an argument.', input);
+  };
+}
+
+//mixin to that adds convenient methods for updating the state of a component
+var stateConvenienceMethods = {
+  state: {
+    'Pat': {
+      'cat': {
+        'name':'Mr. Bigglesworth',
+        'age':10
+      },
+      'dog':{
+        'age':10
+      },
+      'squirrel':true,
+      'pigeon':{
+        'fat':true
+      }
+    },
+    'Dmitri': {
+      'cat':{
+        'age': 12,
+        'name':'Spencer'
+      }
+    },
+    'Gunnari': {
+      'turtles':['Pat', 'Spencer']
+    }
+  },
+
+  setState: function(newState) {
+    this.state = assign({}, this.state, newState);
+  },
+
+  replaceState: function(newState) {
+    this.state = assign({}, newState);
+  },
+
+  // getInitialState: function () {
+  //   return {
+  //     'Pat':true,
+  //     'Dmitri':true,
+  //     'Gunnari':true,
+  //     'Spencer':true
+  //   };
+  // },
+
+  //addState FUNCTION: adds the values of the deepest keys in the passed in object to the corresponding deepest keys in the current state, or throws an error if keys don't match
+  //@param propsToAdd OBJECT: required object argument with deepest keys being numbers or strings to add to the current state
+  //@param callback FUNCTION: optional callback argument that will be executed once setState is completed and the component is re-rendered
+  addState: function (propsToAdd, callback) {
+    //throw error if propsToAdd is undefined or not an object
+    invariantArgCheck(propsToAdd);
+    //make a shallow copy of the current state
+    var currentState = assign({}, this.state);
+    //build new state object
+    var newState = buildNewState(currentState, propsToAdd, function(newState, key, currentStateAtKey, newPropsAtKey) {
+      //throw error if newPropsAtKey is not a number or string
+      invariantNumberOrStringCheck(newPropsAtKey);
+      //perform addition on the corresponding deep keys
+      newState[key] = currentStateAtKey + newPropsAtKey;
+    });
+    //set state with the updated values and a callback (if provided), triggering a re-render
+    this.setState(newState, callback);
+  },
+
+
+  //subtractState FUNCTION: subtracts the value of the deepest keys in the passed in object to the corresponding deepest keys in the current state, or throws an error if keys don't match
+  //@param propsToAdd OBJECT: required object argument with deepest keys being numbers to subtract the current state from
+  //@param callback FUNCTION: optional callback argument that will be executed once setState is completed and the component is re-rendered
+  subtractState: function(propsToSubtract, callback) {
+    //throw error if propsToSubtract is undefined or not an object
+    invariantArgCheck(propsToSubtract);
+    //make a shallow copy of the current state
+    var currentState = assign({}, this.state);
```

Comment:
I don't think we need to invoke `assign` here because we are already making a shallow copy of state at the first line in `buildNewState`.